File: app/src/main/java/com/buddha/mindboard/data/repository/Repository.java
package com.buddha.mindboard.data.repository;

import android.content.Context;
import android.support.annotation.NonNull;


import com.buddha.mindboard.data.datasource.LocalDataSource;
import com.buddha.mindboard.data.datasource.RemoteDataSource;
import com.buddha.mindboard.data.datasource.base.BaseDataSource;
import com.buddha.mindboard.data.datasource.base.DataSource;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;

import static dagger.internal.Preconditions.checkNotNull;

/**
 * Created by Buddha Saikia on 13-11-2018.
 */
@Singleton
public class Repository extends BaseDataSource implements DataSource.Greetings {

    private Context context;
    @NonNull
    private final RemoteDataSource remoteDataSource;
    @NonNull
    private final LocalDataSource localDataSource;

    @Inject
    public Repository(@NonNull Context context,
                      @NonNull RemoteDataSource awRemoteDataSource,
                      @NonNull LocalDataSource awLocalDataSource) {
        this.context = context;
        this.remoteDataSource = checkNotNull(awRemoteDataSource);
        this.localDataSource = checkNotNull(awLocalDataSource);
    }

    @Override
    public Observable<String> greetings() {
        return remoteDataSource.greetings();
    }
}

File: app/src/main/java/com/buddha/mindboard/data/repository/Repository.kt
package com.buddha.mindboard.data.repository

import com.buddha.mindboard.data.datasource.LocalDataSource
import com.buddha.mindboard.data.datasource.RemoteDataSource
import com.buddha.mindboard.data.datasource.base.BaseDataSource
import com.buddha.mindboard.data.datasource.base.DataSource
import com.buddha.mindboard.data.model.Datum
import dagger.internal.Preconditions.checkNotNull
import io.reactivex.Observable
import retrofit2.Response
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Buddha Saikia on 13-11-2018.
 */
@Singleton
class Repository @Inject
constructor(
    remoteDataSource: RemoteDataSource,
    localDataSource: LocalDataSource
) : BaseDataSource(), DataSource {

    private val remoteDataSource: RemoteDataSource = checkNotNull(remoteDataSource)
    private val localDataSource: LocalDataSource = checkNotNull(localDataSource)

    override fun getData(): Observable<Response<List<Datum>>>? {
        return remoteDataSource.getData()
    }
}
